resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/RefillRateLimiter.java

```diff
@@ -0,0 +1,476 @@
+     * Object)} this method will wait for a while before try one more time. This technique was
+     * originally described in this
+     * <a href="https://arxiv.org/abs/1305.5800"> paper</a>
+     * and showed great results with {@link AtomicRateLimiter} in benchmark tests.
+     *
+     * @param current the expected value
+     * @param next    the new value
+     * @return {@code true} if successful. False return indicates that the actual value was not
+     * equal to the expected value.
+     */
+    private boolean compareAndSet(final State current, final State next) {
+        if (state.compareAndSet(current, next)) {
+            return true;
+        }
+        parkNanos(1); // back-off
+        return false;
+    }
+
+
+    /**
+     * A side-effect-free function that can calculate next {@link State} from current. It determines
+     * time duration that you should wait for the given number of permits and reserves it for you,
+     * if you'll be able to wait long enough.
+     *
+     * @param permits        number of permits
+     * @param timeoutInNanos max time that caller can wait for permission in nanoseconds
+     * @param activeState    current state of {@link RefillRateLimiter}
+     * @return next {@link State}
+     */
+    private State calculateNextState(final int permits, final long timeoutInNanos,
+                                     final State activeState) {
+        long nanosSinceLastUpdate = nanosSinceLastUpdate(activeState);
+        int nexPermissions = calculateNextPermissions(activeState, nanosSinceLastUpdate);
+
+        long nextNanosToWait = nanosToWaitForPermission(
+            permits, activeState.config.getNanosPerPermit(), nexPermissions);
+
+        State nextState = reservePermissions(activeState, permits, timeoutInNanos,
+            nexPermissions, nextNanosToWait);
+        return nextState;
+    }
+
+    private int calculateNextPermissions(State activeState, long nanosSinceLastUpdate) {
+        long accumulatedPermissions = accumulatedPermissions(activeState, nanosSinceLastUpdate);
+        return (int) min(activeState.config.getPermitCapacity(), accumulatedPermissions);
+    }
+
+    private long accumulatedPermissions(State activeState, long nanosSinceLastUpdate) {
+        int currentPermissions = activeState.activePermissions;
+        long permissionsBatches = calculateBatches(activeState.config.getNanosPerPermit(), nanosSinceLastUpdate);
+        return permissionsBatches + currentPermissions;
+    }
+
+    private long nanosSinceLastUpdate(State activeState) {
+        return currentNanoTime() - activeState.updatedAt;
+    }
+
+    /**
+     * Calculate the batches, should be inlined
+     * @param nanosPerPermission
+     * @param nanosSinceLastUpdate
+     * @return
+     */
+    private long calculateBatches(long nanosPerPermission, long nanosSinceLastUpdate) {
+        if(nanosPerPermission==0) {
+            return Long.MAX_VALUE;
```

Comment:
In case of 0 we grand infinite permissions?

Reply:
Due to the benchmark tests using a very big number of permissions nanosPerPermission were set to 0, thus caused this edge case to happen. 

More specific 

https://github.com/resilience4j/resilience4j/blob/fed8e7b64f2c54933e00db414fba0dcb20b52c2f/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RateLimiterConfig.java#L234-L237

As seen on the code it is tackled by Giving Infinite permissions. 

Other ways to be tackled would be either through an exception when setting the config and changing the benchmark for this specific rate limiter. Another workaround is to move the config division to calculateNextState (RefillRateLimiter.calculateNextState), thus through some multiplication and division re-arrangements  we avoid the edge case of division by zero. 
It used to be this way but performed less on the benchmarks due to 2 extra calculations.

Reply:
Maybe we should treat such configuration as invalid and require at least nanosPerPermission=1.

Reply:
I proceeded on throwing an IllegalStateException when it comes to the RefillRateLimiterConfig.Builder.build()

`Current settings lead to zero nanos per permission, adjust period and limit`

https://github.com/resilience4j/resilience4j/blob/31a4107469664a6ac00ab80643511180f7655b7f/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RefillRateLimiterConfig.java#L164

and proceeded on throwing an IllegalArgumentException when it comes to constructing a RefillRateLimiterConfig

`At least 1 nanos per permission should be provided`

https://github.com/resilience4j/resilience4j/blob/31a4107469664a6ac00ab80643511180f7655b7f/resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/RefillRateLimiterConfig.java#L41-L44